### ISSUE TITLE
Feedback needed - New feature for the template engine.

### ISSUE THREAD

**Issue by gliechtenstein** [Jasonette/documentation#29]
A few days ago I pushed a new feature for the template engine which fixes one of the most annoying problems people have been facing. 

# Problem
For example here's a problematic situation:

```json
{
  "{{#each $jason.posts}}": {
    "type": "vertical",
    "components": [{
      "type": "label",
      "text": "{{content}}"
    }, {
      "type": "label",
      "text": "{{$jason.username}}"
    }]
  }
}
```

This looks like it should work but it DOES NOT work.

More specifically, the `{{$jason.username}}` part doesn't work.

That's because it's inside the context of `$jason.posts` loop. When you think you're accessing `$jason.username`, you would be actually accessing `$jason.posts[0].$jason.username`, `$jason.posts[1].$jason.username`,etc.  

# Solution

We cannot use above markup because it's ambiguous what you're trying to do. After all, you may actually have an attribute called `$jason` under one of the post items, and may indeed be trying to access `$jason.posts[0].$jason.username`!

To handle this without ambiguity we need a way to **access the root scope from inside the loop**. And that's what this new feature does--it lets you access the root context using the keyword `$root`. Here's the updated JSON, which WILL work in the latest develop branch.

```json
{
  "{{#each $jason.posts}}": {
    "type": "vertical",
    "components": [{
      "type": "label",
      "text": "{{content}}"
    }, {
      "type": "label",
      "text": "{{$root.$jason.username}}"
    }]
  }
}
```

Since `$root` functions as the gateway to access the root context, this means you can access not just `$jason` but any other variables like `$cache` or `$get` this way. Here's an example:

```json
{
  "{{#each $jason.posts}}": {
    "type": "vertical",
    "components": [{
      "type": "label",
      "text": "{{content}}"
    }, {
      "type": "label",
      "text": "{{$root.$get.some_local_variable}}"
    }]
  }
}
```

# Discussion
I haven't made this official so looking for any suggestions on the naming (Right now it's `$root`), or if you have a better idea on how to handle this problem, or any feedback, please feel free to share.

p.s.
BTW I do plan on open sourcing this JS in the future as well.

**Comment by brad** [comment]
@gliechtenstein Maybe this is a silly question, but why can't we make `$jason` the keyword (instead of `$root`) and say you are not allowed to use `$jason` as an object key, as in `$jason.posts[1].$jason.username`. Is there any reason why one would want to use the name `$jason` in child like that?

**Comment by gliechtenstein** [comment]
Hmm, then how would we express the `$jason` objects that are returned from actions as return values?

For example, when we `$render`, the template always starts with

```
{
  "templates": {
    "body": {
      "sections": [{
        "items": {
          "{{#each $jason}}": {
            "type": "label",
            "text": "{{name}}"
          }
        }
      }]
    }
  }
}
```

**Comment by brad** [comment]
@gliechtenstein Ahh, good point. I didn't see the whole picture. I like the idea of introducing `$root` and I'm personally happy with the name. :+1:

**Comment by sg02** [comment]
I guess $ stands for variable, # stands for evaluation, so, '#.variable' might mean a value of a variable, which ' stands for half of a value. In case hierarchy of variable is needed, it may be expressed by . and - just like in typescript.

**Comment by sg02** [comment]
The above proposal is not using any thing related to html.
Perhaps, // is comment in js and might be hidden evaluations. and it's started with //# without any spaces for the first.

Then, the synergy of # and // stand for evaluation and comment like these:

{{ #if }} ==> //#if

{{ $jason.username }} ==> '#.username'

"{{ #each $json.posts }}" : {
  ...
}
==> //#each "$json.posts" : {
  ...
//}

**Comment by sg02** [comment]
Without any of js,

#if

"#.username"

"##each $json.posts##"

# with single word, ## with sentences.
# or ## are just evaluation area.

**Comment by gliechtenstein** [comment]
@sg02 thanks for sharing the idea!

If I understand correctly, you're suggesting we use `#` as the delimiter for any template expressions within a string, right?

One thing I realized while investigating different options is these symbols used to declare the range of a template needs to have a direction. For example `{{` or `<%=` marks the beginning and `}}` or `%>` marks the end.

If we don't have this type of directionality, there are many edge cases the template engine won't be able to handle easily, for example imagine this case:

```
$jason = {
  "sound": "ha"
};
```

And we have a template that looks like this:

```
{
  "message": "I laugh: {{$jason.sound}},{{$jason.sound}},{{$jason.sound}}"
}
```

If we convert this into the `#` syntax,

```
{
  "message": "I laugh: #$jason.sound#,#$jason.sound#,#$jason.sound#"
}
```

As you can see, it's not easy to quickly see what's going on here, unless you figure out the entire nested structure of the string you never know whether `$jason.sound` or `,` in the `#$jason.sound#,#` sequence is the template. And it's not just hard for humans, without a clear direction it becomes harder for the engine to parse since it needs to create a nested structure every time the parsing is done, and there's no need to introduce this complexity when you can simply choose another token with directionality. And I think that's why all template engines have a distinct beginning token and an ending token.

Also another factor is that `#` is a css ID selector. And the template engine sometimes uses jQuery syntax to parse HTML into JSON ([See here](http://docs.jasonette.com/templates/#step-3-use-jquery-syntax-to-parse-and-render)) and this would be problematic.

Basically it all comes down to:

1. The token should not conflict with any JS/CSS keywords
2. Needs to be easy to parse (both for humans and the machines)

---

p.s.
Plot twist: BTW this thread is not about what you were thinking haha, it's about supporting the `$root` context. Maybe I should create a separate issue for the "what token to use for declaring templates"

**Comment by sg02** [comment]
I recommended these 3 types.

'''
1. #reserved_keyword_only

2. #.variable_to_be_a_single_value

3. ##u_may_write_even_ur_long_fuction##
'''

I agree that html template engine was number 3 type only.

And your example is number 2 type like these:

As-was example:

'''
{
  "message": "I laugh: {{$jason.sound}},{{$jason.sound}},{{$jason.sound}}"
}
'''

To-be example:

'''
{
  "message": "I laugh: #.sound,#.sound,#.sound"
}
'''

Please, notice that a single variable will be just a value with '#.'. it means that it's like just js closures seemingly.

PS: As I see, $root would remain in the code but not be used by most users.

**Comment by gliechtenstein** [comment]
How would i do something like this?

```
{
  "message": "I laugh: ha.ha.ha.he"
}
```

Maybe like this?

```
{
  "message": "I laugh: #.sound.#.sound.#.sound.he"
}
```
 
What if `#.sound` actually has an attribute named `#`? Or if it had an attribute called `he`? How do we know if the user intended to access those attributes or if they just meant to print `"#"` ro `"he"`?

**Comment by sg02** [comment]
Well, how about <code>^#</code> or <code>#^</code> as an escape character to avoid js conflict?

And it's no problem at all with <code>he</code> I guess. Because <code>#.single_word</code> is literally a single word.

**Comment by gliechtenstein** [comment]
> Well, how about ^# or #^ as an escape character to avoid js conflict?

Haha yes I've been asking you these questions not because I disagree with you completely but because this was exactly the thought process I went through--I had the same idea (just use a hashtag!) and had tried to make it work--and what started out as a clean solution started looking weird as I tried to support complex situations like these (I actually tried rewriting some markups into this notation `#^ ... ^#` and it ended up looking horrible compared to the `{{ }}`, which was the opposite of what was intended).

> Because #.single_word is literally a single word.

I'm not sure if I understood correctly but why is it just a single word? it can be an array or an object too. For example:

```
{
  "items": {{$jason.items}}"
}
```

can be parsed into something like 

```
{
  "items": ["a", "b", "c"]
}
```

**Comment by sg02** [comment]
Someone meant to print <code>#</code>, it might be written as <code>^##</code>.
Because the escape character is <code>^#</code>. (Easy to find <code>#</code> first and after.)

Your example will be like:
```
{
  "items": #.items
}
```
and I will change the definition a bit like:
```
2. #.single_variable_to_be_a_single_line_value_by_preprocessing
```
you can notice that <code>#.single_word</code> is still itself.

In case, to expand single_word to hierachy_word, it may be changed <code>.</code> to <code>-</code> and it would be more complex and might be postpone to implementing.

**Comment by gliechtenstein** [comment]
Thanks for sharing the idea, I'll take it into account. 

Actually, reading your comment and thinking about it more gave me some good ideas. Trying it out right now. Thanks!

**Comment by sg02** [comment]
In case, $ is used as tokens for templates,
I will change my recommendations like these:

```
1. ##:reserved_keyword

2. ##.single_variable

3. ###sentence_like_functions###
```

Only `#` alone is not proper for uri, html id, c language, etc.
